### PR TITLE
download media later fix #254

### LINKED ISSUE #254: Download media later
Hello! I have a doubt. Is there a possibility to download a media afterwards? That is, by the message id can I retrieve it and download it?

### PR DESCRIPTION
**Pull request by AxeemHaider**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (tuyuribr) [inline .vscode/settings.json:1]: Delete this file

Comment (AxeemHaider) [reply .vscode/settings.json]: Done

**Comment by PurpShell** [comment]
Just what I thought of, will do a proper review later. 👍

File: index.d.ts [hunk 2]
```diff
@@ -115,6 +115,9 @@ declare namespace WAWebJS {
         /** Searches for messages */
         searchMessages(query: string, options?: { chatId?: string, page?: number, limit?: number }): Promise<Message[]>
 
+        /** Download message media for cached messages only */
+        downloadMedia(mediaInfo: MediaInfo): Promise<MessageMedia>
```
Comment (PurpShell) [inline index.d.ts:119]:
```suggestion
        downloadMedia(): Promise<MessageMedia>
```

File: src/Client.js [hunk 3]
```diff
@@ -586,6 +586,71 @@ class Client extends EventEmitter {
...
+     * @param {string} [mediaInfo.encFilehash]
+     * @param {string} [mediaInfo.filehash]
+     * @param {string} [mediaInfo.mediaKey]
+     * @param {number} [mediaInfo.mediaKeyTimestamp]
+     * @param {string} [mediaInfo.type]
+     * @param {string} [mediaInfo.mimetype]
+     * @param {string} [mediaInfo.filename]
+     * @returns {Promise<MessageMedia>}
+     */
+    async downloadMedia(mediaInfo) {
```
Comment (PurpShell) [inline src/Client.js:603]:
```suggestion
    async downloadMedia() {
```

File: src/Client.js [hunk 4]
```diff
@@ -586,6 +586,71 @@ class Client extends EventEmitter {
...
+        const result = await this.pupPage.evaluate(async (mediaInfo) => {
+
+            const msg = window.Store.Msg.get(mediaInfo.serializedId);
```
Comment (PurpShell) [inline src/Client.js:606]:
```suggestion
            const msg = window.Store.Msg.get(this.id._serialized);
```

File: src/Client.js [hunk 5]
```diff
@@ -586,6 +586,71 @@ class Client extends EventEmitter {
...
+     * @param {Object} [mediaInfo]
+     * @param {string} [mediaInfo.serializedId]
+     * @param {string} [mediaInfo.directPath]
+     * @param {string} [mediaInfo.encFilehash]
+     * @param {string} [mediaInfo.filehash]
+     * @param {string} [mediaInfo.mediaKey]
+     * @param {number} [mediaInfo.mediaKeyTimestamp]
+     * @param {string} [mediaInfo.type]
+     * @param {string} [mediaInfo.mimetype]
+     * @param {string} [mediaInfo.filename]
```
Comment (PurpShell) [inline src/Client.js:600]:
```suggestion
```

File: src/Client.js [hunk 6]
```diff
@@ -586,6 +586,71 @@ class Client extends EventEmitter {
...
+                return {
+                    data,
+                    mimetype: mediaInfo.mimetype,
+                    filename: mediaInfo.filename
+                };
+            } catch (e) {
+                if(e.status && e.status === 404) return undefined;
+                throw e;
+            }
+        }, mediaInfo);
```
Comment (PurpShell) [inline src/Client.js:648]:
```suggestion
        }, this.mediaInfo);
```

**Review by purpshell — CHANGES_REQUESTED** [review]
some changes to make this seamless :+1:

File: index.d.ts (hunk 2 above)
Comment (AxeemHaider) [reply index.d.ts:119]: Why you removed the `mediaInfo` and how you are accessing it like `this.mediaInfo` if you look it carefully this `downloadMedia()` method is in `Client.js` which does not contain any information about messages.

Comment (PurpShell) [reply index.d.ts:119]: > Why you removed the `mediaInfo` and how you are accessing it like `this.mediaInfo` if you look it carefully this `downloadMedia()` method is in `Client.js` which does not contain any information about messages.

huh? let me re-review

**Comment by PurpShell** [comment]
@AxeemHaider Rather than have a "Client.downloadMedia", could we have a  fix to `Message.downloadMedia`??

You could create a new message or use #847

**Comment by AxeemHaider** [comment]
Currently we are not sure we can create `new Message` every time as mention in this (comment)[https://github.com/pedroslopez/whatsapp-web.js/pull/847#issuecomment-907903145] If we are able to create new Message then we don't need to save `mediaInfo` Simply if you can get Message later than you can also get media later, than there is no need for this PR.

**Comment by AxeemHaider** [comment]
I thing right now it's better to provide `downloadMedia` method in `Client`

**Pull request closed without merging**